### PR TITLE
fix BAM/SAM header setting when writing on cluster

### LINKED ISSUE #676: BAMHeader not set when running on a cluster
When running `transform` with `.bam` file as the output, I see the following error:
```
Exception in thread "main" java.lang.AssertionError: assertion failed: Cannot return header if not attached.
        at scala.Predef$.assert(Predef.scala:179)
        at org.bdgenomics.adam.rdd.read.ADAMBAMOutputFormat$.getHeader(ADAMBAMOutputFormat.scala:60)
        at org.bdgenomics.adam.rdd.read.ADAMBAMOutputFormat.<init>(ADAMBAMOutputFormat.scala:68)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at java.lang.Class.newInstance(Class.java:379)
        at org.apache.spark.rdd.InstrumentedOutputFormat.<init>(InstrumentedOutputFormat.scala:33)
        at org.bdgenomics.adam.rdd.read.InstrumentedADAMBAMOutputFormat.<init>(ADAMBAMOutputFormat.scala:71)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at java.lang.Class.newInstance(Class.java:379)
        at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopDataset(PairRDDFunctions.scala:958)
```

I tried to fix this with a similar strategy used by @fnothaft for the VCF header (see https://github.com/bigdatagenomics/adam/compare/master...hammerlab:bam-save-header?expand=1), but did not seem to take.  Any ideas on this?

### PR DESCRIPTION
**Pull request by ryan-williams**
fixes #676

### CODE REVIEW

**Comment by ryan-williams** [comment]
test failure should be fixed. Jenkins, retest this please?

**Comment by ryan-williams** [comment]
Jenkins, retest this please

File: adam-core/src/main/scala/org/bdgenomics/adam/rich/RichAlignmentRecord.scala [hunk 1]
```diff
@@ -68,7 +68,7 @@ class IlluminaOptics(val tile: Long, val x: Long, val y: Long) {}
 
 class RichAlignmentRecord(val record: AlignmentRecord) {
 
-  lazy val referenceLength: Int = RichAlignmentRecord.referenceLengthFromCigar(record.getCigar.toString)
+  lazy val referenceLength: Int = RichAlignmentRecord.referenceLengthFromCigar(record.getCigar)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rich/RichAlignmentRecord.scala]: Thanks for getting these; we've got several redundant `toString`s hanging around from the pre-[bdg-formats/36](https://github.com/bigdatagenomics/bdg-formats/pull/36) merged.

**Comment by ryan-williams** [comment]
rebased and squashed

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
[Another head-scratcher](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/731/SCALAVER=2.10,hadoop.version=2.3.0,label=centos/console):



```
FATAL: Could not checkout null with start point 8bb41727766ea4ca8c51379f77036be0ca64701e
```

**Comment by fnothaft** [comment]
Jenkins, retest this please.

@ryan-williams that's a known issue in the Jenkins Github Pull Request Builder plugin. It occasionally fails to check out the PR after a force push to the branch.

**Comment by ryan-williams** [comment]
ah, good to know! I'll try to hold off the force-push-squash until after I've gotten a green build in the future.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
Merged! Thanks @ryan-williams!

**Comment by ryan-williams** [comment]
nice! thank you, @fnothaft!

**Pull request merged**